File: app/src/main/java/app/grapheneos/pdfviewer/fragment/DocumentPropertiesFragment.java
package app.grapheneos.pdfviewer.fragment;

import android.app.Activity;
import android.app.Dialog;
import android.os.Bundle;
import android.widget.ArrayAdapter;

import androidx.annotation.NonNull;
import androidx.fragment.app.DialogFragment;

import com.google.android.material.dialog.MaterialAlertDialogBuilder;

import app.grapheneos.pdfviewer.R;

import java.util.ArrayList;
import java.util.List;

public class DocumentPropertiesFragment extends DialogFragment {
    public static final String TAG = "DocumentPropertiesFragment";

    private static final String KEY_DOCUMENT_PROPERTIES = "document_properties";

    private List<String> mDocumentProperties;

    public static DocumentPropertiesFragment newInstance(final List<CharSequence> metaData) {
        final DocumentPropertiesFragment fragment = new DocumentPropertiesFragment();
        final Bundle args = new Bundle();

        args.putCharSequenceArrayList(KEY_DOCUMENT_PROPERTIES, (ArrayList<CharSequence>) metaData);
        fragment.setArguments(args);

        return fragment;
    }

    @Override
    public void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);

        if (getArguments() != null) {
            mDocumentProperties = getArguments().getStringArrayList(KEY_DOCUMENT_PROPERTIES);
        }
    }

    @NonNull
    @Override
    public Dialog onCreateDialog(Bundle savedInstanceState) {
        final Activity activity = requireActivity();
        final MaterialAlertDialogBuilder dialog = new MaterialAlertDialogBuilder(activity)
                .setPositiveButton(android.R.string.ok, null);

        if (mDocumentProperties != null) {
            dialog.setTitle(getString(R.string.action_view_document_properties));
            dialog.setAdapter(new ArrayAdapter<>(activity, android.R.layout.simple_list_item_1,
                    mDocumentProperties), null);
        } else {
            dialog.setTitle(R.string.document_properties_retrieval_failed);
        }
        return dialog.create();
    }
}

File: app/src/main/java/app/grapheneos/pdfviewer/fragment/DocumentPropertiesFragment.kt
package app.grapheneos.pdfviewer.fragment

import android.app.Dialog
import android.graphics.Typeface
import android.os.Bundle
import android.text.SpannableStringBuilder
import android.text.Spanned
import android.text.style.StyleSpan
import android.widget.ArrayAdapter
import androidx.fragment.app.DialogFragment
import androidx.fragment.app.activityViewModels
import app.grapheneos.pdfviewer.R
import app.grapheneos.pdfviewer.loader.DocumentProperty
import app.grapheneos.pdfviewer.viewModel.PdfViewModel
import com.google.android.material.dialog.MaterialAlertDialogBuilder

class DocumentPropertiesFragment : DialogFragment() {

    private val viewModel by activityViewModels<PdfViewModel>()

    private fun formatProperties(properties: Map<DocumentProperty, String>): List<CharSequence> {
        return properties.map { (property, value) ->
            val name = getString(property.nameResource)
            SpannableStringBuilder()
                .append(name)
                .append(":\n")
                .append(value)
                .apply {
                    setSpan(
                        StyleSpan(Typeface.BOLD),
                        0,
                        name.length,
                        Spanned.SPAN_EXCLUSIVE_EXCLUSIVE
                    )
                }
        }
    }

    override fun onCreateDialog(savedInstanceState: Bundle?): Dialog {
        val properties = viewModel.documentProperties.value

        return MaterialAlertDialogBuilder(requireActivity())
            .setPositiveButton(android.R.string.ok, null).apply {
                if (!properties.isNullOrEmpty()) {
                    setTitle(getString(R.string.action_view_document_properties))
                    setAdapter(
                        ArrayAdapter(
                            requireActivity(),
                            android.R.layout.simple_list_item_1,
                            formatProperties(properties)
                        ), null
                    )
                } else {
                    setTitle(R.string.document_properties_retrieval_failed)
                }
            }
            .create()
    }

    companion object {

        const val TAG = "DocumentPropertiesFragment"

        @JvmStatic
        fun newInstance() = DocumentPropertiesFragment()
    }
}
